modules/ROOT/pages/dataweave-cookbook-decode-encode.adoc

```diff
@@ -0,0 +1,158 @@
+. Run the project in Studio.
+. Load the `0.0.0.0:8081/transform` into a browser.
++
+This action creates a PDF file in the specified location and generates a separate `transform` file with the message in the Set Payload component, `PDF file created in the directory specified in the Write operation.`
+
+[[ex01]]
+== Encode a PDF File to Base64
+
+You can simulate the transformation in the Read operation from the <<xml_config, XML configuration>>. This example performs Base64 encoding on a PDF file. More specifically, it converts a Base64 file stream (`application/octet-stream`) to Binary and outputs the Binary in the `text/plain` format.
+
+The example uses the following functions:
+
+*  `readUrl` from the `Core (dw::Core)` module to input a PDF file.
++
+The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`.
+* `fromBase64` from the `dw::core::Binaries` module to transform a Base64 string to a binary value.
+
+//NOTE that we can't yet automate this example. Outputs text format.
+//Need to check if PDF input is supported by validator.
+.DataWeave Script:
+[source,dataweave,linenums]
+----
+%dw 2.0
+import * from dw::core::Binaries
+var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
+output text/plain
+---
+toBase64(myPDF as Binary)
+----
+
+.Output (partial results):
+[source,json,linenums]
+----
+JVBERi0xLjYNJeLjz9MNCjM3IDAgb2JqIDw8L0xpbmVhcml6ZWQgMS9MIDIwNTk3L08gNDAvRSAx
+...
+----
+
+
+[[ex02]]
+== Decode Base64
+
+You can simulate the transformation in the Write operation from the <<xml_config, XML configuration>>. This DataWeave example converts a Base64 file stream (`application/octet-stream`) to Binary and outputs the Binary in the `text/plain` format. It does not generate a PDF file; a Write operation is required to generate such a file.
+
+The examples use the following functions from the `dw::core::Binaries` module:
+
+*  `readUrl` to input a PDF file.
++
+The example assumes that the PDF file is in the `src/main/resources` directory of a Mule project in Anypoint Studio. It provides the name of the PDF file and specifies the MIME type as `application/octet-stream`.
+* `fromBase64` in the `dw::core::Binaries` module to transform a Base64 string to a binary value.
+
+The script reads the input PDF as an octet-stream MIME type. From within the call to `fromBase64`, the script coerces that input to a string value before coercing the string to Base64.
+
+//NOTE that we can't yet automate this example. Outputs text format.
+//Need to check if PDF input is supported by validator.
+.DataWeave Script:
+[source,dataweave,linenums]
+----
+%dw 2.0
+import * from dw::core::Binaries
+var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
+var myPDFasBinary = toBase64(myPDF as Binary)
+output text/plain
```

Comment:
This is confusing, shouldn't it be binary too as in the other example? In a mule app this would fail

Reply:
That seemed to work in Studio (no error thrown). `readUrl` doesn't accept `"application/pdf with binary"` as the second arg but is ok with `"binary"`.

Reply:
Modified the `fromBase64` example with the suggested change, replacing `output text/plain` with `output application/pdf with binary`

Reply:
I expected that, too, but I got those results when I ran this in the Transform Message component:
```
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "application/octet-stream")
//var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
//output application/octet-stream
---
fromBase64(payload as String) as Binary
```

I put in a logger (with fx set to `"testing output" ++ payload`)  after that TM component, saved, stopped, and restarted the app, and got this result  in the console:
```
testing output%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
0000001386 00000 n
```
same went for the DW script in a standalone TM component (not in that flow).
Script:
```
%dw 2.0
import * from dw::core::Binaries
var myPDF = readUrl("classpath://pdf-test.pdf", "binary")
var myPDFasBinary = toBase64(myPDF as Binary)
output application/PDF with binary
---
fromBase64(myPDFasBinary as String) as Binary
```
returned

```
%PDF-1.6
%����
37 0 obj <</Linearized 1/L 20597/O 40/E 14115/N 1/T 19795/H [ 1005 215]>>
endobj
                 
xref
37 34
0000000016 00000 n
```